crates/shim-sgx/src/handler/mod.rs

```diff
@@ -179,15 +181,60 @@ impl guest::Handler for Handler<'_> {
         Ok(())
     }
 
-    // Until EDMM, we can't change any page permissions.
-    // What you get is what you get. Fake success.
     fn mprotect(
         &mut self,
         _platform: &impl Platform,
-        _addr: NonNull<c_void>,
-        _len: c_size_t,
-        _prot: c_int,
+        addr: NonNull<c_void>,
+        length: c_size_t,
+        prot: c_int,
     ) -> sallyport::Result<()> {
+        let addr = addr.as_ptr() as usize;
+        let length = (length + 0xfff) & !0xfff;
+
+        if addr & 0xfff != 0 || length == 0 {
+            return Err(EINVAL);
+        }
+
+        // Given that the hardware asserts on !W^R, the function, by practical
+        // means, only RWX is disregarded from non-empty subsets of permissions.
+        if prot != PROT_READ && prot != (PROT_READ | PROT_WRITE) && prot != (PROT_READ | PROT_EXEC)
+        {
+            debugln!(self, "({:>#016x}, {:>#016x}, {:>#02x})", addr, length, prot);
+            return Err(EACCES);
+        }
+
+        let mut heap = HEAP.write();
+
+        if heap.contains(Region::new(Address::new(addr), Address::new(addr + length))) == None {
+            return Err(ENOMEM);
+        }
+
+        // TODO: Change the backend to zero the permissions instead of
+        // restricting them to 'read' (SGX2 series v4 issue).
+        self.reset_sgx_permissions(NonNull::new(addr as *mut _).unwrap(), length)?;
+
+        for i in 0..length / Page::SIZE {
+            let virt_addr = VirtAddr::new((addr + i * Page::SIZE) as u64);
+            let page_addr = PageAddr::from_start_address(virt_addr).unwrap();
```

Comment:
Can we avoid `unwrap()`?

Reply:
The failure condition can never happen, as `virt_addr` is always page aligned. I would consider this more like an assert.

Reply:
@npmccallum there is unsafe version of `from_start_address()`: https://docs.rs/x86_64/latest/x86_64/structures/paging/page/struct.Page.html#method.from_start_address_unchecked

Reply:
With this I'm open for suggestion but do not have anything better at hand. See the Safety-section of [from_start_address_unchecked](https://docs.rs/x86_64/latest/x86_64/structures/paging/page/struct.Page.html#method.from_start_address_unchecked):

[Safety](https://docs.rs/x86_64/latest/x86_64/structures/paging/page/struct.Page.html#safety)
The address must be correctly aligned.

[source](https://docs.rs/x86_64/latest/src/x86_64/structures/paging/page.rs.html#96-101)

Reply:
I'll write a helper function for this conversion. It's complex enough and repeats multiple places.

Reply:
done